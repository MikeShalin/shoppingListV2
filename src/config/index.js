/** @flow **/
export const key = 'productList'
