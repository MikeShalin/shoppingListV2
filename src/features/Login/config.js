/** @flow **/

export const formFields = {
  login: {
    label: 'First Name',
    placeholder: 'First Name',
  },
  password: {
    label: 'Enter Password',
    placeholder: 'Enter Password',
  },
}