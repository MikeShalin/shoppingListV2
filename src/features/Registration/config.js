/** @flow **/

export const formFields = {
  login: {
    label: 'Enter email',
    placeholder: 'Enter email',
  },
  password: {
    label: 'Enter Password',
    placeholder: 'Enter Password',
  },
}