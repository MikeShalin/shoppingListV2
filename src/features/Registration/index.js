/** @flow **/
import * as React from 'react'
import { Fragment } from 'react'
import {
  compose,
  withProps,
  defaultProps,
} from 'recompose'
import { inject, observer } from 'mobx-react/index'
import { Message } from 'semantic-ui-react'

import Form from 'Core/Form'
import {
  password as passwordConfig,
  login as loginConfig,
  withTerms as withTermsConfig,
  message,
} from './config'

type TProps = {
  registration: {
    login: string,
    password: string,
    withTerms: boolean,
    success: boolean,
    error: boolean,
    onChange: (value: {}) => boolean,
    onSubmit: () => boolean,
    fetching: boolean,
  },
  handlerChange: (value: {}) => void,
  header: string,
  content: string,
}

//TODO подумать как лучше передать пропсы и не нарушает ли это какой нибудь принцип?
const FormExampleForm = ({
                           registration: {
                             login,
                             password,
                             withTerms,
                             onChange,
                             onSubmit,
                             fetching,
                             ...registration
                           },
                           handlerChange,
                           header,
                           content,
                         }: TProps) => (
  <Fragment>
    <Form
      login={{ onChange: handlerChange, ...loginConfig }}
      password={{ onChange: handlerChange, ...passwordConfig }}
      withTerms={{
        onChange: () => onChange({ withTerms: !withTerms }),
        checked: withTerms,
        ...withTermsConfig,
      }}
      onSubmit={onSubmit}
      loading={fetching}
      {...registration}
    >
      <Message
        header={header}
        content={content}
        {...registration}
      />
    </Form>
  </Fragment>
)

const composed = compose(
  inject('registration'),
  defaultProps({
    successMessage: message.success,
    errorMessage: message.error,
  }),
  withProps(({ registration: { onChange } }) => ({
    handlerChange: (e) => {
      const {
        currentTarget: {
          name,
          value,
        },
      } = e
      onChange({ [name]: value })
    },
  })),
  observer,
  withProps(({
               registration: {
                 success,
               },
               successMessage,
               errorMessage,
             }) => ({
    header: success
      ? successMessage.header
      : errorMessage.header,
    content: success
      ? successMessage.content
      : errorMessage.content,
  })),
  observer,
)

export default composed(FormExampleForm)