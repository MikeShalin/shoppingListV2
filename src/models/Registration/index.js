import { types } from 'mobx-state-tree'
import _ from 'lodash'

import { getItem, setItem } from 'Utils/localStorage'
import { key } from 'Config'

//TODO разделить на api и т.д.
const Registration = types
  .model('Store', {
    login: types.string,
    password: types.string,
    withTerms: types.boolean,
    success: types.boolean,
    error: types.boolean,
    fetching: types.boolean,
  })
  .actions(self => ({
    onChange(target) {
      const [[key, value]] = Object.entries(target)
      self[key] = value
    },
    onSuccess(){
      self.success = true
      self.error = false
      self.setFetching(false)
    },
    onError(){
      self.success = false
      self.error = true
      self.setFetching(false)
    },
    setFetching(state) {
      self.fetching = state
    },
    onSubmit() {
      const { login, password, withTerms } = self
      self.setFetching(true)
      if (!login || !password || !withTerms) {

        /** имитация загрузки **/
        //TODO calApi
        _.delay(self.onError , 1000)
        return false
      }
      self.login = ''
      self.password = ''
      self.withTerms = false

      //TODO не сбрасывается текст полей при отправке
      //TODO текст предупреждения
      //TODO Field Error
      //TODO Disabled Field submit, пока все не введешь
      //TODO Required
      /** имитация загрузки **/
      _.delay(self.onSuccess , 1000)
      return true
    },
  }))

export default Registration