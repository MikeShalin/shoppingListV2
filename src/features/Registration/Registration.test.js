import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { formFields } from './config'
import Registration from './index'

describe('config Registration component', () => {
  it('formFields matches initial state', () => {
    expect(formFields).toEqual({
      login: {
        label: 'Enter email',
        placeholder: 'Enter email',
      },
      password: {
        label: 'Enter Password',
        placeholder: 'Enter Password',
      },
    })
  })
})

describe('Registration component', () => {
  it('matches early snapshots', () => {
    const tree = shallow(<Registration />)
    expect(shallowToJson(tree)).toMatchSnapshot()
  })
})