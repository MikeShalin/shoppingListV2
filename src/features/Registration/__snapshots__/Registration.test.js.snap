// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration component matches early snapshots 1`] = `
<MainForm
  login={
    Object {
      "label": "Enter email",
      "placeholder": "Enter email",
    }
  }
  password={
    Object {
      "label": "Enter Password",
      "placeholder": "Enter Password",
    }
  }
/>
`;
